### PR TITLE
[XPU] remove cast for elementwise_add(float, bf16/fp16)

### PR DESCRIPTION
**Pull request by dynamicheart**
### PR Category

Custom Device

### PR Types

Performance

### Description

- remove cast for elementwise_add(float, bf16/fp16)
- unittest for this PR has been added in this PR: https://github.com/PaddlePaddle/Paddle/pull/63865

Reference PR:
- https://github.com/PaddlePaddle/Paddle/pull/63293
- https://github.com/PaddlePaddle/Paddle/pull/63842/files#diff-bfa7dc1ccc441c58f8baf7af0e7407b4e9b4dfb9b4f0c2da652d385e7a2f7e59R143

### CODE REVIEW

File: paddle/phi/kernels/xpu/elementwise_add_kernel.cc [hunk 1]
```diff
@@ -37,18 +37,44 @@ void AddKernel(const Context& dev_ctx,
...
+        ret = xpu::add_mul_type<float, XPUYType, float>(
+            dev_ctx.x_context(), x_data, y_data, z_data, x.numel());
+      } else {
+        using YType = DataTypeToCppType<phi::DataType::FLOAT16>::type;
+        using XPUYType = typename XPUTypeTrait<YType>::Type;
+        auto y_data = reinterpret_cast<const XPUYType*>(y.data<YType>());
+        ret = xpu::add_mul_type<float, XPUYType, float>(
+            dev_ctx.x_context(), x_data, y_data, z_data, x.numel());
+      }
+      PADDLE_ENFORCE_XDNN_SUCCESS(ret, "add_mul_type");
```
Comment (Aurelius84) [inline paddle/phi/kernels/xpu/elementwise_add_kernel.cc:63]: 这里的报错信息支持common::errors::XX 类型提示么？如果支持的话，建议加一下报错类型

Comment (dynamicheart) [reply paddle/phi/kernels/xpu/elementwise_add_kernel.cc:63]: 好的， 我研究下

File: paddle/phi/common/type_promotion.h [hunk 2]
```diff
@@ -145,6 +145,11 @@ inline bool NeedTypePromotion(const std::string& op_name,
         (y == phi::DataType::BFLOAT16 || y == phi::DataType::FLOAT16)) {
       return false;
     }
+#elif defined(PADDLE_WITH_XPU)
+    if ((op_name == "add" || op_name == "add_") && x == DataType::FLOAT32 &&
```
Comment (dynamicheart) [inline paddle/phi/common/type_promotion.h:149]: @Aurelius84 @zxcd 麻烦看看CUDA版本是否是漏了add_算子，这样XPU和CUDA可以合并，add_(fp32， bf16/fp16)最一开始的需求是来自于main_grad的inplace_add：https://github.com/PaddlePaddle/Paddle/pull/54415/files#diff-1c8fea91efa1fa5f00f4b766fb11723120b74ec26365724cca4be8a0792946d9R62

File: paddle/phi/kernels/xpu/elementwise_add_kernel.cc (hunk 1 above)
Comment (dynamicheart) [reply paddle/phi/kernels/xpu/elementwise_add_kernel.cc:63]: 目前原本的XPUElementwise里面也是调用的`PADDLE_ENFORCE_XDNN_SUCCESS`，包括其它XPU算子也是用的`PADDLE_ENFORCE_XDNN_SUCCESS`，至于`common::errors::XX`报错信息的建议后续可以专门批量去改，看看是否能加到PADDLE_ENFORCE_XDNN_SUCCESS宏里面

**Review by zhangyk0314 — APPROVED** [review]
LGTM

**Review by qili93 — APPROVED** [review]
LGTM

**Review by cqulilujia — APPROVED** [review]
LGTM

**Pull request merged**